### ISSUE TITLE
error processing query messages upon subcription with geo-filter

### ISSUE THREAD

**Issue by fgalan** [telefonicaid/fiware-orion#3729]
Sometimes we get error in traces like this one:

```
time=2020-11-15T23:29:32.260Z | lvl=ERROR | corr=6a081632-279a-11eb-b0cb-0242ac11001a | trans=1605470146-001-00000001337 | from=10.95.235.101 | srv=cb00802cep | subsrv=/sscb00802cep | comp=Orion | op=AlarmManager.cpp[211]:dbError | msg=Raising alarm DatabaseError: nextSafe(): { $err: "error processing query: ns=orion-cb00802cep.entitiesTree: $and
    _id.servicePath $eq "/sscb00802cep"
    attrNames $eq "position"
    GEONEAR  field...", code: 291, ok: 0.0 }
```

The following request seems to trigger the failure:

```
POST /v2/subscriptions, request payload (293 bytes): {"subject":{"entities":[{"idPattern":".*"}],"condition":{"attrs":["temperature"],"expression":{"georel":"near;maxDistance:1000","geometry":"point","coords":"40.418889,-3.691944"}}},"notification":{"http":{"url":"http://10.95.190.130:9090/notices"},"attrsFormat":"legacy","attrs":["position"]}}, response code: 201
POST /v2/subscriptions, request payload (293 bytes): {"subject":{"entities":[{"idPattern":".*"}],"condition":{"attrs":["temperature"],"expression":{"georel":"near;maxDistance:1000","geometry":"point","coords":"40.418889,-3.691944"}}},"notification":{"http":{"url":"http://10.95.190.130:9090/notices"},"attrsFormat":"legacy","attrs":["position"]}}, response code: 201
POST /v2/subscriptions, request payload (293 bytes): {"subject":{"entities":[{"idPattern":".*"}],"condition":{"attrs":["temperature"],"expression":{"georel":"near;maxDistance:1000","geometry":"point","coords":"40.418889,-3.691944"}}},"notification":{"http":{"url":"http://10.95.190.130:9090/notices"},"attrsFormat":"legacy","attrs":["position"]}}, response code: 201
POST /v2/subscriptions, request payload (293 bytes): {"subject":{"entities":[{"idPattern":".*"}],"condition":{"attrs":["temperature"],"expression":{"georel":"near;maxDistance:1000","geometry":"point","coords":"40.418889,-3.691944"}}},"notification":{"http":{"url":"http://10.95.190.130:9090/notices"},"attrsFormat":"legacy","attrs":["position"]}}, response code: 201
POST /v2/subscriptions, request payload (293 bytes): {"subject":{"entities":[{"idPattern":".*"}],"condition":{"attrs":["temperature"],"expression":{"georel":"near;maxDistance:1000","geometry":"point","coords":"40.418889,-3.691944"}}},"notification":{"http":{"url":"http://10.95.190.130:9090/notices"},"attrsFormat":"legacy","attrs":["position"]}}, response code: 201
POST /v2/subscriptions, request payload (293 bytes): {"subject":{"entities":[{"idPattern":".*"}],"condition":{"attrs":["temperature"],"expression":{"georel":"near;maxDistance:1000","geometry":"point","coords":"40.418889,-3.691944"}}},"notification":{"http":{"url":"http://10.95.190.130:9090/notices"},"attrsFormat":"legacy","attrs":["position"]}}, response code: 201
POST /v2/subscriptions, request payload (293 bytes): {"subject":{"entities":[{"idPattern":".*"}],"condition":{"attrs":["temperature"],"expression":{"georel":"near;maxDistance:1000","geometry":"point","coords":"40.418889,-3.691944"}}},"notification":{"http":{"url":"http://10.95.190.130:9090/notices"},"attrsFormat":"legacy","attrs":["position"]}}, response code: 201
POST /v2/subscriptions, request payload (268 bytes): {"subject":{"entities":[{"idPattern":".*"}],"condition":{"attrs":["temperature"],"expression":{"georel":"near;maxDistance:1000","geometry":"point","coords":"-80.190,25.774"}}},"notification":{"http":{"url":"http://10.95.190.130:10031/notify"},"attrs":["temperature"]}}, response code: 201
POST /v2/subscriptions, request payload (265 bytes): {"subject":{"entities":[{"idPattern":".*"}],"condition":{"attrs":["temperature"],"expression":{"georel":"near;minDistance:0","geometry":"point","coords":"-80.190,25.774"}}},"notification":{"http":{"url":"http://10.95.190.130:10031/notify"},"attrs":["temperature"]}}, response code: 201
```

**Comment by fgalan** [comment]
Useful curls for debuging. They correspond to first, before last and last.

```
curl localhost:1026/v2/subscriptions -s -S -H 'Content-Type: application/json' -d @- <<EOF
{
	"subject": {
		"entities": [{
			"idPattern": ".*"
		}],
		"condition": {
			"attrs": ["temperature"],
			"expression": {
				"georel": "near;maxDistance:1000",
				"geometry": "point",
				"coords": "40.418889,-3.691944"
			}
		}
	},
	"notification": {
		"http": {
			"url": "http://10.95.190.130:9090/notices"
		},
		"attrsFormat": "legacy",
		"attrs": ["position"]
	}
}
EOF

#2

curl localhost:1026/v2/subscriptions -s -S -H 'Content-Type: application/json' -d @- <<EOF
{
	"subject": {
		"entities": [{
			"idPattern": ".*"
		}],
		"condition": {
			"attrs": ["temperature"],
			"expression": {
				"georel": "near;maxDistance:1000",
				"geometry": "point",
				"coords": "-80.190,25.774"
			}
		}
	},
	"notification": {
		"http": {
			"url": "http://10.95.190.130:10031/notify"
		},
		"attrs": ["temperature"]
	}
}
EOF

#3

curl localhost:1026/v2/subscriptions -s -S -H 'Content-Type: application/json' -d @- <<EOF
{
	"subject": {
		"entities": [{
			"idPattern": ".*"
		}],
		"condition": {
			"attrs": ["temperature"],
			"expression": {
				"georel": "near;minDistance:0",
				"geometry": "point",
				"coords": "-80.190,25.774"
			}
		}
	},
	"notification": {
		"http": {
			"url": "http://10.95.190.130:10031/notify"
		},
		"attrs": ["temperature"]
	}
}
EOF
```

**Comment by fgalan** [comment]
We have found that under normal conditions, the error doesn't take place. Steps to reproduce:

1. Start Orion
2. Run the following in `mongo orion` shell: `db.entities.dropIndex({"location.coords" : "2dsphere"})`
3. Do request query and error appears in the logs

Alternatively:

1. Start Orion
2. Remove entities collection
3. Re-created with `db.createCollection("entities")` (it starts without any index)
3. Do request query and error appears in the logs

The error disappears once the geo-index is created again, which can be due to:

* A new entity is created
* Orion is restarted (it check geo-index at startup time and creates it if doesn't exist)

The index shouldn't disappear under normal utilization of Orion. Oncly out-of-band operations could do that, so probably it's a good idea that Orion warn about it with an ERROR message in the logs in these cases.

**Comment by fgalan** [comment]
Some additional findings:

_One_

The issue will happen with **entities collection exist without the geo-index**. However, it will not happen if the **entities collection doesn't exist at all**. The second is the usual situation when a given DB service doesn't exist and it is created with a new subscription. I mean, let's assume `orion-foo` DB doesn't exist previously and the following request is recevied:

```
curl localhost:1026/v2/subscriptions -H 'fiware-service: foo' -H 'fiware-servicepath: /SSS' -s -S -H 'Content-Type: application/json' -d @- <<EOF
< some subscription using expression with geo-filter> 
EOF
```

That will lead to this in DB:

```
> use orion-foo
switched to db orion-foo
> show collections
csubs
```

and the ERROR trace will not occur.

Bottom line: the creation of new DBs upon service first use in `-multiservice` cases will not cause the issue. The hypothesis that the only situation that can cause an ERROR trace here is due to out-of-band DB manipulation stills.

_Two_

This issue doesn't append if `options=skipInitialNotification` is used. Makes sense: in this case no query is done in entities collection to look for entities covered by the subscription, so no error is raised due to the lack of geo-index.

**Comment by fgalan** [comment]
Although this is not an issue for Orion, I think it can be an issue in ORCH. See: https://github.com/telefonicaid/orchestrator/issues/301
